### PR TITLE
Support gateway api BackendTLSPolicy

### PR DESCRIPTION
**Pull request by flawedmatrix**
The BackendTLSPolicy will allow a user to connect an httproute to a backend service with TLS.

Implements #5929

- Only implements BackendTLSPolicy for HTTPRoute
- Only allows Services for spec.targetRef
- Allows ConfigMap or Secret for spec.TLS.CACertRefs
- Adds backendtlspolicies to the ClusterRole for Contour
- Adds configmaps to the ClusterRole for Contour
- Controller reconciles on BackendTLSPolicy and ConfigMaps now
- BackendTLSPolicy spec.targetRef can specify SectionName to be port name of a service to target a particular section of the service.

In the interest of keeping this size of this change relatively small, we opted not to implement all of the GEP (https://gateway-api.sigs.k8s.io/geps/gep-1897/)

Specifically, we have not yet implemented BackendTLSPolicy for TLSRoute or GRPCRoute, precedence handling for BackendTLSPolicy (described here: https://gateway-api.sigs.k8s.io/geps/gep-713/#conflict-resolution), displaying status conditions on the BackendTLSPolicy, among others we may have missed.

We have also opted not to implement `wellKnownCACerts` since it was an implementation-specific detail of the GEP (https://github.com/kubernetes-sigs/gateway-api/issues/2726) and right now support for System certs was not supported anywhere in Contour.

### CODE REVIEW

**Comment by flawedmatrix** [comment]
cc @christianang

File: cmd/contour/serve.go
Comment (sunjayBhatia) [inline cmd/contour/serve.go:642]: i think we could only start this informer if the backendtlspolicy feature is enabled, could augment the `features` map below to be `map[string][]string` so if a feature is disabled/enabled we can enable/disable any other required informers/controllers

**Comment by sunjayBhatia** [comment]
An additional follow up issue will be that the Gateway Provisioner will need to set up the appropriate RBAC for Contour instances it provisions so that BackendTLSPolicy can be used

File: internal/dag/cache_test.go
Comment (sunjayBhatia) [inline internal/dag/cache_test.go:1094]: nit
```suggestion
		"insert backendtlspolicy targeting backend Service": {
```

File: internal/dag/gatewayapi_processor.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:85]: looks like this needs to be set in `cmd/serve.go` so general upstream TLS settings can flow through

we can add some follow up e2e tests for these parameters (tls version etc.)

Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor.go:85]: looks like it is set in `featuretests` but not `cmd/serve.go` (which unfortunately only gets covered by e2es)

File: internal/dag/cache_test.go
Comment (sunjayBhatia) [inline internal/dag/cache_test.go:390]: should be able to add a case for a Secret referenced by a BackendTLSPolicy (on the remove case as well)

File: internal/dag/cache_test.go
Comment (sunjayBhatia) [inline internal/dag/cache_test.go:3013]: i think there should be some consideration of what namespace the backendtlspolicy resides in as well, at least as an initial pass let's not allow policies that live in a different namespace from the Service (for the purposes of this test, the targetRef target namespace) they reference

so we may need to always construct a targetRef with a namespace which in the processor this would be the namespace of the HTTPRoute if not specified on the backendref, otherwise the ns of the backendref

File: internal/dag/gatewayapi_processor.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:2000]: this may not be specified if no cross-ns reference is present, so likely have to make this the ns of the HTTPRoute if empty

**Comment by sunjayBhatia** [comment]
Looks like this file does not include BackendTLSPolicy: https://github.com/kubernetes-sigs/gateway-api/blob/v1.0.0/config/crd/experimental/kustomization.yaml 🙃

**Comment by sunjayBhatia** [comment]
> Looks like this file does not include BackendTLSPolicy: https://github.com/kubernetes-sigs/gateway-api/blob/v1.0.0/config/crd/experimental/kustomization.yaml 🙃
> 
> so we never automatically picked it up in our examples:
> 
> https://github.com/projectcontour/contour/blob/9eb2838cac25a37a2424b0b5b11323724162b176/hack/generate-gateway-yaml.sh#L17

we could update this instead to pull from the release artifact: https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/experimental-install.yaml

**Comment by christianang** [comment]
> > Looks like this file does not include BackendTLSPolicy: https://github.com/kubernetes-sigs/gateway-api/blob/v1.0.0/config/crd/experimental/kustomization.yaml 🙃
> > so we never automatically picked it up in our examples:
> > https://github.com/projectcontour/contour/blob/9eb2838cac25a37a2424b0b5b11323724162b176/hack/generate-gateway-yaml.sh#L17
> 
> we could update this instead to pull from the release artifact: https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/experimental-install.yaml

Yeah we noticed that. That is interesting that BackendTLSPolicy is in the released artifact. We figured it might make sense to just PR the kustomization.yaml, but I'm fine with switching to the released artifact if that makes more sense.

**Comment by sunjayBhatia** [comment]
> > > Looks like this file does not include BackendTLSPolicy: https://github.com/kubernetes-sigs/gateway-api/blob/v1.0.0/config/crd/experimental/kustomization.yaml 🙃
> > > so we never automatically picked it up in our examples:
> > > https://github.com/projectcontour/contour/blob/9eb2838cac25a37a2424b0b5b11323724162b176/hack/generate-gateway-yaml.sh#L17
> > 
> > 
> > we could update this instead to pull from the release artifact: https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/experimental-install.yaml
> 
> Yeah we noticed that. That is interesting that BackendTLSPolicy is in the released artifact. We figured it might make sense to just PR the kustomization.yaml, but I'm fine with switching to the released artifact if that makes more sense.

yeah we'll have to probably switch to downloading the released artifact since the old tag won't be updated

made a PR to fix this here: https://github.com/kubernetes-sigs/gateway-api/pull/2736

File: cmd/contour/serve.go
Comment (christianang) [reply cmd/contour/serve.go:642]: Not sure I follow the change for features to go from `map[string]struct{}` to `map[string][]string`. What would go into the new string array value type here? I see the serve.go currently deletes from the map if a feature is disabled. Is that not enough?

File: internal/dag/gatewayapi_processor.go
Comment (christianang) [reply internal/dag/gatewayapi_processor.go]: Good catch. Done.

File: internal/dag/cache_test.go
Comment (christianang) [reply internal/dag/cache_test.go:3085]: Yeah logic here has some hidden complexity. I did a bit of a rewrite, now matching the namespace provided on the targetref to the backendtlspolicy namespace instead, then check that the backendtlspolicy namespace matches the spec.targetRef.namespace afterward. Also added some logic around default namespaces and empty namespaces e.g if spec.targetRef.namespace is empty that should be fine since that should be used for cross namespace-ing, which is unsupported. Added additional testing here as well, but could always use another look to check if anything is missed or unclear.

File: internal/dag/cache_test.go
Comment (christianang) [reply internal/dag/cache_test.go:421]: Done. It should be noted that the secretRetriggerBuild function always returns true on ca certs, which is what is always provided to backendtlspolicies, so no changes to the implementation was done here, but testing it is still valuable just to provide intent in case the logic here changes.

File: internal/dag/gatewayapi_processor.go
Comment (christianang) [reply internal/dag/gatewayapi_processor.go:85]: Good catch. I added the upstreamTLS to the processor in `cmd/serve.go`, and we can follow up with UpstreamTLS testing at the e2e level in a follow up PR if we want to do that as suggested.

File: internal/dag/cache_test.go
Comment (christianang) [reply internal/dag/cache_test.go]: Done.

File: cmd/contour/serve.go
Comment (sunjayBhatia) [reply cmd/contour/serve.go:642]: the map key would be feature names, which currently are only resource kinds we inform on, and then the string list could be kinds that a particular feature/top level kind depends on, in this case backendtlspolicy also needs a configmap informer to work fully, so if you disable backendtlspolicy you dont need to have a configmap informer etc.

it might be overkill, tbh since we also have an if statement checking if backendtlspolicy is enabled, we can add the creation of the configmap informer in there

Comment (christianang) [reply cmd/contour/serve.go]: Ended up moving the informOnResource line down into the feature check for backendltspolicy.

**Comment by christianang** [comment]
@sunjayBhatia made some changes from your feedback and got the checks passing. I think this is ready for another review.

File: internal/dag/gatewayapi_processor.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:2019]: we could add some coverage for this stuff but really it is covered by the `LookupBackendTLSPolicyByTargetRef` helper tbh, so maybe we can leave it or even drop this check?

File: internal/dag/gatewayapi_processor.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:2029]: we could add some negative unit tests for the rest of these and/or wait for the status for BackendTLSPolicy implementation to cover these errors

might be good at least to add a single test that demonstrates that if there are errors in this area (Policy references a secret that doesn't exist etc.) then there won't be any upstream TLS on the Cluster

**Review by sunjayBhatia — COMMENTED** [review]
Besides some v minor things I think this is basically good to go. Before merging too would be good to get some issues created for all the outstanding things for follow-ups:
- Supporting BackendTLS for GRPCRoute
- Supporting BackendTLS for TLSRoute
- Status Condition logic for BackendTLSPolicies 
- Interaction between the existing Service annotation for specifying backend protocol and BackendTLSPolicy
- Handling conflict resolution if there are *multiple Policies targeting a Service*
- e2e testing upstream TLS settings w/ Gateway API Routes (to cover the Gateway processor being wired up properly in serve.go)
- look at cluster name generation to see if we need to include all of the ca cert secret names in the hashed name
- and any other things that we can think of

File: internal/dag/builder_test.go
Comment (izturn) [inline internal/dag/builder_test.go:4285]: Shall we define some helper functions to create the objects?, this files is too large now

**Comment by christianang** [comment]
> Besides some v minor things I think this is basically good to go. Before merging too would be good to get some issues created for all the outstanding things for follow-ups:
> 
> * Supporting BackendTLS for GRPCRoute
> * Supporting BackendTLS for TLSRoute
> * Status Condition logic for BackendTLSPolicies
> * Interaction between the existing Service annotation for specifying backend protocol and BackendTLSPolicy
> * Handling conflict resolution if there are _multiple Policies targeting a Service_
> * e2e testing upstream TLS settings w/ Gateway API Routes (to cover the Gateway processor being wired up properly in serve.go)
> * look at cluster name generation to see if we need to include all of the ca cert secret names in the hashed name
> * and any other things that we can think of

Created follow up issues for the above. The only other follow up item I could think of is cross namespace certificate secrets with ReferenceGrants.

Happy to add more issues as more follow up items come up.

File: internal/dag/gatewayapi_processor.go
Comment (christianang) [reply internal/dag/gatewayapi_processor.go:2019]: I think it makes sense to drop the check since you are right it is covered by the helper.

File: internal/dag/builder_test.go
Comment (christianang) [reply internal/dag/builder_test.go]: That makes sense. Done.

File: internal/dag/gatewayapi_processor.go
Comment (christianang) [reply internal/dag/gatewayapi_processor.go:2025]: Added some negative tests, which did happen to catch something wrong in the implementation 😓 , so that is good. When we add conditions the expectations of these tests can probably change to account for that.

File: test/e2e/gateway/backend_tls_policy_test.go
Comment (christianang) [reply test/e2e/gateway/backend_tls_policy_test.go:181]: Makes sense. Done.

File: internal/envoy/cluster.go
Comment (christianang) [reply internal/envoy/cluster.go:50]: Broke out the follow issue to follow up: https://github.com/projectcontour/contour/issues/6141

**Comment by christianang** [comment]
I addressed the latest round of feedback.

**Review by sunjayBhatia — APPROVED** [review]
LGTM, just a merge conflict to resolve

cc @projectcontour/maintainers if you want to take a look as well

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1059]: You'll need to add this as a valid option for the `--disable-feature` flag, defined above in this file (`serve.Flag("disable-feature", ...`)

Also, since the rest of the values are plural, let's do the same here -- `backendtlspolicies`.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1091]:
```suggestion
		// Create and register the BackendTLSPolicy controller with the manager.
```

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1098]: `internal/k8s/helpers.go#IsObjectEqual` should be updated to include `BackendTLSPolicy`, it should be able to be handled in the same case as the rest of the Gateway API objects.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1094]: `internal/k8s/helpers.go#IsObjectEqual` should be updated to include `ConfigMap`, it looks like it will need a custom comparison, probably similar to Secrets?

**Review by skriss — CHANGES_REQUESTED** [review]
Couple initial comments, still working through DAG logic

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:1094]: I'm a little concerned about increased memory usage from the informer caching all ConfigMaps in the cluster. For Secrets, we did the following to avoid storing all the data from irrelevant Secrets: https://github.com/projectcontour/contour/blob/29201bbd13e5cf3477e35a8744f53bb13befb3b5/cmd/contour/serve.go#L253-L258

I think we should probably do something similar here for ConfigMaps since the vast majority of them will not be relevant for Contour. Probably checking for the `ca.crt` key, like you're doing in the DAG cache?

**Review by skriss — CHANGES_REQUESTED** [review]

File: cmd/contour/serve.go
Comment (christianang) [reply cmd/contour/serve.go]: Done.

File: cmd/contour/serve.go
Comment (christianang) [reply cmd/contour/serve.go]: Good catch. Done.

File: cmd/contour/serve.go
Comment (christianang) [reply cmd/contour/serve.go:1120]: Done.

File: cmd/contour/serve.go
Comment (christianang) [reply cmd/contour/serve.go:1116]: Done.

Comment (christianang) [reply cmd/contour/serve.go:1116]: Makes sense. Done.

**Comment by christianang** [comment]
Thanks for the review Steve. I've updated the PR to address your comments.

**Comment by skriss** [comment]
Oops, looks like `TestIsEqualFallback` was using ConfigMap as its "unsupported" kind, but it's now "supported" in IsObjectEqual 😄 https://github.com/projectcontour/contour/actions/runs/7731651205/job/21079863873?pr=6119#step:6:75

**Comment by christianang** [comment]
Ah missed running those tests locally. I updated the tests to account for the ConfigMap changes.

**Review by skriss — APPROVED** [review]
LGTM, will leave in case @sunjayBhatia wanted to take another look

**Review by sunjayBhatia — APPROVED** [review]
New changes LGTM!

**Pull request merged**
